lib/workers/repository/update/pr/changelog/source.ts

```diff
@@ -0,0 +1,271 @@
+import URL from 'node:url';
+import is from '@sindresorhus/is';
+import { logger } from '../../../../../logger';
+import { getPkgReleases } from '../../../../../modules/datasource';
+import type { Release } from '../../../../../modules/datasource/types';
+import * as allVersioning from '../../../../../modules/versioning';
+import * as packageCache from '../../../../../util/cache/package';
+import { regEx } from '../../../../../util/regex';
+import { trimSlashes } from '../../../../../util/url';
+import type { BranchUpgradeConfig } from '../../../../types';
+import { slugifyUrl } from './common';
+import { addReleaseNotes } from './release-notes';
+import { getInRangeReleases } from './releases';
+import type {
+  ChangeLogError,
+  ChangeLogRelease,
+  ChangeLogResult,
+} from './types';
+
+export abstract class ChangeLogSource {
+  private platform;
+  private datasource;
+  private cacheNamespace: string;
+
+  constructor(
+    platform: 'bitbucket' | 'github' | 'gitlab',
+    datasource: 'bitbucket-tags' | 'github-tags' | 'gitlab-tags'
+  ) {
+    this.platform = platform;
+    this.datasource = datasource;
+    this.cacheNamespace = `changelog-${platform}-release`;
+  }
+
+  abstract getCompareURL(
+    baseUrl: string,
+    repository: string,
+    prevHead: string,
+    nextHead: string
+  ): string;
+
+  abstract getAPIBaseUrl(config: BranchUpgradeConfig): string;
+
+  async getAllTags(endpoint: string, repository: string): Promise<string[]> {
+    const tags = (
+      await getPkgReleases({
+        datasource: this.datasource,
+        packageName: repository,
+        versioning:
+          'regex:(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?',
+      })
+    )?.releases;
+
+    if (is.nullOrUndefined(tags) || is.emptyArray(tags)) {
+      logger.debug(
+        `No ${this.datasource} tags found for repository: ${repository}`
+      );
+
+      return [];
+    }
+
+    return tags.map(({ version }) => version);
+  }
+
+  public async getChangeLogJSON(
+    config: BranchUpgradeConfig
+  ): Promise<ChangeLogResult | null> {
+    logger.trace(`getChangeLogJSON for ${this.platform}`);
+
+    const versioning = config.versioning!;
+    const currentVersion = config.currentVersion!;
+    const newVersion = config.newVersion!;
+    const sourceUrl = config.sourceUrl!;
+    const packageName = config.packageName!;
+    const sourceDirectory = config.sourceDirectory!;
```

Comment:
`sourceUrl` is used in the existing logic as part of the cache key, so I'm assuming this can never be null.

I've removed the `!` from `sourceDirectory`